[ICE]: `...has parameters, but no args were provided in instantiate`

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
#![feature(min_generic_const_args)]

trait A<T> {}
trait Trait<const N: usize> {}

impl A<[usize; fn_item]> for () {}

fn fn_item(_: impl Trait<usize>) {}

fn main() {}

```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.98.0-nightly (bc2112ed5 2026-06-18)
binary: rustc
commit-hash: bc2112ed56c99fa649e09ab3ab286afab3d9059a
commit-date: 2026-06-18
host: x86_64-unknown-linux-gnu
release: 1.98.0-nightly
LLVM version: 22.1.7
```

### Error output

```
<output>
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
warning: the feature `min_generic_const_args` is incomplete and may not be safe to use and/or cause compiler crashes
 --> /tmp/crash.rs:1:12
  |
1 | #![feature(min_generic_const_args)]
  |            ^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #132980 <https://github.com/rust-lang/rust/issues/132980> for more information
  = note: `#[warn(incomplete_features)]` on by default


thread 'rustc' (1293383) panicked at /rustc-dev/bc2112ed56c99fa649e09ab3ab286afab3d9059a/compiler/rustc_type_ir/src/binder.rs:636:13:
<() as A<[usize; fn_item::<impl Trait<usize>>]>> has parameters, but no args were provided in instantiate
stack backtrace:
   0:     0x7efcbabfd889 - <<std[9155b05baf46f0b5]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[b26e7c5dab710dcd]::fmt::Display>::fmt
   1:     0x7efcbb20ebde - core[b26e7c5dab710dcd]::fmt::write
   2:     0x7efcbac13cb6 - <std[9155b05baf46f0b5]::sys::stdio::unix::Stderr as std[9155b05baf46f0b5]::io::Write>::write_fmt
   3:     0x7efcbabd2b7e - std[9155b05baf46f0b5]::panicking::default_hook::{closure#0}
   4:     0x7efcbabf0573 - std[9155b05baf46f0b5]::panicking::default_hook
   5:     0x7efcb9c49fc1 - std[9155b05baf46f0b5]::panicking::update_hook::<alloc[4a86d35f348a3b04]::boxed::Box<rustc_driver_impl[b791fc15675dcea2]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7efcbabf0852 - std[9155b05baf46f0b5]::panicking::panic_with_hook
   7:     0x7efcbabd2c42 - std[9155b05baf46f0b5]::panicking::panic_handler::{closure#0}
   8:     0x7efcbabc7169 - std[9155b05baf46f0b5]::sys::backtrace::__rust_end_short_backtrace::<std[9155b05baf46f0b5]::panicking::panic_handler::{closure#0}, !>
   9:     0x7efcbabd46ad - __rustc[b4c4dac14d05b73]::rust_begin_unwind
  10:     0x7efcb7b0455c - core[b26e7c5dab710dcd]::panicking::panic_fmt
  11:     0x7efcb74f7cfd - <rustc_trait_selection[747f355ebc30f854]::traits::select::SelectionContext>::match_impl::{closure#0}
  12:     0x7efcbb2820f4 - <rustc_trait_selection[747f355ebc30f854]::traits::select::SelectionContext>::assemble_candidates_from_impls
  13:     0x7efcbb2855f9 - <rustc_trait_selection[747f355ebc30f854]::traits::select::SelectionContext>::candidate_from_obligation_no_cache
  14:     0x7efcbb7ac89e - <rustc_trait_selection[747f355ebc30f854]::traits::select::SelectionContext>::poly_select::{closure#0}
  15:     0x7efcbb9472ca - <rustc_trait_selection[747f355ebc30f854]::traits::fulfill::FulfillProcessor as rustc_data_structures[788feccef7b56953]::obligation_forest::ObligationProcessor>::process_obligation
  16:     0x7efcbb205b51 - <rustc_data_structures[788feccef7b56953]::obligation_forest::ObligationForest<rustc_trait_selection[747f355ebc30f854]::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection[747f355ebc30f854]::traits::fulfill::FulfillProcessor>
  17:     0x7efcbb7e8565 - <rustc_trait_selection[747f355ebc30f854]::traits::fulfill::FulfillmentContext<rustc_trait_selection[747f355ebc30f854]::traits::FulfillmentError> as rustc_infer[d147af7b5b190580]::traits::engine::TraitEngine<rustc_trait_selection[747f355ebc30f854]::traits::FulfillmentError>>::evaluate_obligations_error_on_ambiguity
  18:     0x7efcbb7e6496 - rustc_hir_analysis[f8649c0bc5b32aaa]::check::wfcheck::check_impl
  19:     0x7efcbc0a82bd - rustc_hir_analysis[f8649c0bc5b32aaa]::check::check::check_item_type
  20:     0x7efcbc09bab2 - rustc_hir_analysis[f8649c0bc5b32aaa]::check::wfcheck::check_well_formed
  21:     0x7efcbc09ba93 - rustc_query_impl[f814d044a8094e2a]::query_impl::check_well_formed::invoke_provider_fn::__rust_begin_short_backtrace
  22:     0x7efcbc09b072 - rustc_query_impl[f814d044a8094e2a]::execution::try_execute_query::<rustc_data_structures[788feccef7b56953]::vec_cache::VecCache<rustc_span[4065efe3f3bc8177]::def_id::LocalDefId, rustc_middle[cb7c4bb37f51aac1]::query::erase::ErasedData<[u8; 1usize]>, rustc_middle[cb7c4bb37f51aac1]::dep_graph::graph::DepNodeIndex>, false>
  23:     0x7efcbc09ae07 - rustc_query_impl[f814d044a8094e2a]::query_impl::check_well_formed::execute_query_non_incr::__rust_end_short_backtrace
  24:     0x7efcbc0978bb - rustc_hir_analysis[f8649c0bc5b32aaa]::check::wfcheck::check_type_wf
  25:     0x7efcbc0977c9 - rustc_query_impl[f814d044a8094e2a]::query_impl::check_type_wf::invoke_provider_fn::__rust_begin_short_backtrace
  26:     0x7efcbc38138d - rustc_query_impl[f814d044a8094e2a]::execution::try_execute_query::<rustc_middle[cb7c4bb37f51aac1]::query::caches::SingleCache<rustc_middle[cb7c4bb37f51aac1]::query::erase::ErasedData<[u8; 1usize]>>, false>
  27:     0x7efcbc381169 - rustc_query_impl[f814d044a8094e2a]::query_impl::check_type_wf::execute_query_non_incr::__rust_end_short_backtrace
  28:     0x7efcbb6a781e - rustc_hir_analysis[f8649c0bc5b32aaa]::check_crate
  29:     0x7efcbb6a9d0c - rustc_interface[84e7048b04034882]::passes::analysis
  30:     0x7efcbc37f906 - rustc_query_impl[f814d044a8094e2a]::execution::try_execute_query::<rustc_middle[cb7c4bb37f51aac1]::query::caches::SingleCache<rustc_middle[cb7c4bb37f51aac1]::query::erase::ErasedData<[u8; 0usize]>>, false>
  31:     0x7efcbc37f5e9 - rustc_query_impl[f814d044a8094e2a]::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
  32:     0x7efcbc399c52 - rustc_interface[84e7048b04034882]::interface::run_compiler::<(), rustc_driver_impl[b791fc15675dcea2]::run_compiler::{closure#0}>::{closure#1}
  33:     0x7efcbc3ceffa - std[9155b05baf46f0b5]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[84e7048b04034882]::util::run_in_thread_with_globals<rustc_interface[84e7048b04034882]::util::run_in_thread_pool_with_globals<rustc_interface[84e7048b04034882]::interface::run_compiler<(), rustc_driver_impl[b791fc15675dcea2]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  34:     0x7efcbc3cedad - <std[9155b05baf46f0b5]::thread::lifecycle::spawn_unchecked<rustc_interface[84e7048b04034882]::util::run_in_thread_with_globals<rustc_interface[84e7048b04034882]::util::run_in_thread_pool_with_globals<rustc_interface[84e7048b04034882]::interface::run_compiler<(), rustc_driver_impl[b791fc15675dcea2]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[b26e7c5dab710dcd]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  35:     0x7efcbc3c5290 - <std[9155b05baf46f0b5]::sys::thread::unix::Thread>::new::thread_start
  36:     0x7efcb5e981b9 - <unknown>
  37:     0x7efcb5f1d21c - <unknown>
  38:                0x0 - <unknown>

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/tmp/rustc-ice-2026-06-19T20_08_42-1293373.txt` to your bug report

note: rustc 1.98.0-nightly (bc2112ed5 2026-06-18) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [check_well_formed] checking that `<impl at /tmp/crash.rs:6:1: 6:32>` is well-formed
#1 [check_type_wf] checking that types are well-formed
#2 [analysis] running analysis passes on crate `crash`
end of query stack
error[E0747]: type provided when a constant was expected
 --> /tmp/crash.rs:8:26
  |
8 | fn fn_item(_: impl Trait<usize>) {}
  |                          ^^^^^
  |
help: if this generic argument was intended as a const parameter, surround it with braces
  |
8 | fn fn_item(_: impl Trait<{ usize }>) {}
  |                          +       +

error: aborting due to 1 previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0747`.
```

</p>
</details>


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_hir_analysis/src/hir_ty_lowering/generics.rs
+++ b/compiler/rustc_hir_analysis/src/hir_ty_lowering/generics.rs
@@ -6,7 +6,7 @@ use rustc_errors::{
 };
 use rustc_hir::def::{DefKind, Res};
 use rustc_hir::def_id::DefId;
-use rustc_hir::{self as hir, DelegationInfo, GenericArg};
+use rustc_hir::{self as hir, GenericArg};
 use rustc_middle::ty::{
     self, GenericArgsRef, GenericParamDef, GenericParamDefKind, IsSuggestable, Ty,
 };
@@ -431,15 +431,12 @@ pub(crate) fn check_generic_arg_count(
     }
 
     let tcx = cx.tcx();
-    let parent_def = tcx.hir_get_parent_item(seg.hir_id).def_id;
 
     // Suppress this warning for delegations as it is compiler generated and lifetimes are
     // propagated while late-bound lifetimes may be present.
-    let explicit_late_bound = match tcx.hir_opt_delegation_info(parent_def) {
-        Some(DelegationInfo { child_seg_id, .. }) if seg.hir_id == *child_seg_id => {
-            ExplicitLateBound::No
-        }
-        _ => prohibit_explicit_late_bound_lifetimes(cx, gen_params, gen_args, gen_pos),
+    let explicit_late_bound = match tcx.hir_is_delegation_child_segment(seg) {
+        true => ExplicitLateBound::No,
+        false => prohibit_explicit_late_bound_lifetimes(cx, gen_params, gen_args, gen_pos),
     };
 
     let mut invalid_args = vec![];

--- a/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs
+++ b/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs
@@ -716,6 +716,7 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
             generic_args: &'a GenericArgs<'tcx>,
             span: Span,
             infer_args: bool,
+            create_synth_args: bool,
             incorrect_args: &'a Result<(), GenericArgCountMismatch>,
         }
 
@@ -828,7 +829,7 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
                                 .instantiate(tcx, preceding_args)
                                 .skip_norm_wip()
                                 .into()
-                        } else if synthetic {
+                        } else if self.create_synth_args && synthetic {
                             Ty::new_param(tcx, param.index, param.name).into()
                         } else if infer_args {
                             self.lowerer.ty_infer(Some(param), self.span).into()
@@ -868,6 +869,7 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
             span,
             generic_args: segment.args(),
             infer_args: segment.infer_args,
+            create_synth_args: tcx.hir_is_delegation_child_segment(segment),
             incorrect_args: &arg_count.correct,
         };
 

--- a/compiler/rustc_middle/src/hir/map.rs
+++ b/compiler/rustc_middle/src/hir/map.rs
@@ -879,6 +879,13 @@ impl<'tcx> TyCtxt<'tcx> {
         self.hir_opt_delegation_info(delegation_id).expect("processing delegation")
     }
 
+    pub fn hir_is_delegation_child_segment(self, segment: &PathSegment<'_>) -> bool {
+        let parent_def = self.hir_get_parent_item(segment.hir_id).def_id;
+
+        self.hir_opt_delegation_info(parent_def)
+            .is_some_and(|info| info.child_seg_id == segment.hir_id)
+    }
+
     #[inline]
     fn hir_opt_ident(self, id: HirId) -> Option<Ident> {
         match self.hir_node(id) {

--- a/tests/ui/const-generics/mgca/synth-gen-arg-ice-158152.rs
+++ b/tests/ui/const-generics/mgca/synth-gen-arg-ice-158152.rs
@@ -1,0 +1,11 @@
+#![feature(min_generic_const_args)]
+
+trait A<T> {}
+trait Trait<const N: usize> {}
+
+impl A<[usize; fn_item]> for () {}
+
+fn fn_item(_: impl Trait<usize>) {}
+//~^ ERROR: type provided when a constant was expected
+
+fn main() {}

--- a/tests/ui/const-generics/mgca/synth-gen-arg-ice-158152.stderr
+++ b/tests/ui/const-generics/mgca/synth-gen-arg-ice-158152.stderr
@@ -1,0 +1,14 @@
+error[E0747]: type provided when a constant was expected
+  --> $DIR/synth-gen-arg-ice-158152.rs:8:26
+   |
+LL | fn fn_item(_: impl Trait<usize>) {}
+   |                          ^^^^^
+   |
+help: if this generic argument was intended as a const parameter, surround it with braces
+   |
+LL | fn fn_item(_: impl Trait<{ usize }>) {}
+   |                          +       +
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0747`.

